---
title: '01 Prep - `r unlist(strsplit(getwd(), "/"))[7]` - `r unlist(strsplit(getwd(), "/"))[8]`'
author:
  - name: "Emir Turkes [emir.turkes@eturkes.com]"
  - name: "UK Dementia Research Institute at UCL"
date: '`r strftime(Sys.time(), "%B %d, %Y")`'
bibliography: '../../../../`r unlist(strsplit(getwd(), "/"))[4]`.bib'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding, output_file = file.path(
      "..", "..", "..",
      "..", "results", unlist(strsplit(getwd(), "/"))[6],
      unlist(strsplit(getwd(), "/"))[7], unlist(strsplit(getwd(), "/"))[8], "01_prep.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
.tocify-subheader .tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 35px; text-indent: 0;}
div.main-container {max-width: none; width: 95%;}
</style>

*This file is a part of the [Pathology Signatures Project](https://github.com/eturkes/pathology-signatures).*

In this document we prepare the gene count matrix for downstream analysis.
The data here is derived from @`r unlist(strsplit(getwd(), "/"))[7]` and will be referenced using the name ``r unlist(strsplit(getwd(), "/"))[7]``.
This analysis focuses on both NFT-bearing (neurofibrillary tangle) and normal neurons from the dataset.

```{r}
# Some standard boilerplate.
# --------------------------
#    This file is part of pathology-signatures.
#    Copyright (C) 2022  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

packages <- c(
  "conflicted", "Seurat", "SingleCellExperiment", "biomaRt", "scuttle", "DT", "parallel", "ggplot2", "dplyr", "ggrepel",
  "viridis"
)
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))
source(file.path("..", "..", "..", "utils.R"))

protocol <- c("human", "droplet", "single-cell", "umis") # See `cluster_pipeline` in `utils.R`.
vars_to_regress <- NULL # See `cluster_pipeline` in `utils.R`.
parallel_override <- NULL # See `parallel_plan` in `utils.R`.

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", 4)
names(metadata_to_plot) <- c("seurat_clusters", "Sample.ID", "SORT", "donor_id")
metadata_to_plot$Sample.ID <- "no_label"
metadata_to_plot$SORT <- "no_label"
metadata_to_plot$donor_id <- "no_label"
# -------------------------------------------------------------------------

analysis_no <- 1
assets_dir <- file.path("..", "..", "..", "..", "assets") # Backed up files and data.
results_dir <- file.path(
  "..", "..", "..",
  "..", "results", unlist(strsplit(getwd(), "/"))[6],
  unlist(strsplit(getwd(), "/"))[7], unlist(strsplit(getwd(), "/"))[8], paste0("0", analysis_no)
)

cache_dir <- file.path(
  "..", "..", "..",
  "..", "cache", unlist(strsplit(getwd(), "/"))[6],
  unlist(strsplit(getwd(), "/"))[7], unlist(strsplit(getwd(), "/"))[8], paste0("0", analysis_no)
)
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 300)
# --------------------------
```

# Cleaning

We start by gathering together the data and forming a SingleCellExperiment (SCE) object.

```{r}
data <- readRDS(file.path(assets_dir, "datasets", "otero-garcia-2022-ssRNAseq", "excitatory_neurons.rds"))
data <- data[ , data$Braak == "VI"]

sce <- as.SingleCellExperiment(data)
rm(data)
sce

batch <- sce$donor_id
```

# QC

First, we add cell QC metric metadata to the SCE object.

```{r}
mart <- useEnsembl("ensembl", "hsapiens_gene_ensembl")
attributes <- c("ensembl_gene_id", "chromosome_name")
gene_anno <- getBM(attributes, "ensembl_gene_id", rownames(sce), mart)
gene_anno <- gene_anno[gene_anno$chromosome_name == "MT", ]

mito <- which(rownames(sce) %in% gene_anno$ensembl_gene_id)

if (length(mito) != 0) {
  sce <- addPerCellQC(sce, list(mito = mito))
  if (is.null(batch)) {
    remove <- quickPerCellQC(colData(sce), percent_subsets = "subsets_mito_percent")
  } else {
    remove <- quickPerCellQC(colData(sce), percent_subsets = "subsets_mito_percent", batch = batch)
  }

  if (protocol[3] == "single-nuc") {
    cat(paste0("Removed ", length(mito), " mitochondrial genes:"), rownames(sce)[mito])
    sce <- sce[-mito, ]
    dim(sce)
  } else {
    cat(
      paste0("Found but did not remove ", length(mito), " mitochondrial genes:"),
      rownames(sce)[mito]
    )
  }

} else {
  sce <- addPerCellQC(sce)
  if (is.null(batch)) {
    remove <- quickPerCellQC(colData(sce))
  } else {
    remove <- quickPerCellQC(colData(sce), batch = batch)
  }
}

names(colData(sce))[(length(names(colData(sce))) - 5):length(names(colData(sce)))]
```

We use adaptive thresholds to consider cells for removal that are outliers by more than 3 MADs.
However, since QC was already performed on this dataset, we do not remove them.

```{r}
sce$discard <- remove$discard
datatable_download(t(colSums(as.matrix(remove))))
```

Finally we add feature QC metrics and remove features not expressed in any cell.

```{r, fig.height = 5}
sce <- addPerFeatureQC(sce)
names(rowData(sce))
dim(sce)

par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
 log10(rowData(sce)$mean + 1e-6), col = "grey80",  main = "",
 breaks = 40, xlab = "log10(Mean Counts Per Gene + 1e-6)")
hist(
 log10((rowData(sce)$detected * dim(sce)[2]) + 1), col = "grey80", main = "",
 breaks = 40, xlab = "log10(Number of Cells Expressing Gene + 1)")
plot(
 log10(rowData(sce)$mean + 1e-6), pch = 16,
 col = rgb(0, 0, 0, 0.4), log10((rowData(sce)$detected * dim(sce)[2]) + 1),
 xlab = "log10(Mean Counts Per Gene + 1e-6)", ylab = "log10(Number of Cells Expressing Gene + 1)")

sce <- sce[rowSums(counts(sce) > 0) > 0, ]
rowData(sce) <- NULL
sce <- addPerFeatureQC(sce)
dim(sce)

par(mfrow = c(1, 3), mar = c(5, 4, 1, 1))
hist(
 log10(rowData(sce)$mean + 1e-6), col = "grey80",  main = "",
 breaks = 40, xlab = "log10(Mean Counts Per Gene + 1e-6)")
hist(
 log10((rowData(sce)$detected * dim(sce)[2]) + 1), col = "grey80", main = "",
 breaks = 40, xlab = "log10(Number of Cells Expressing Gene + 1)")
plot(
 log10(rowData(sce)$mean + 1e-6), pch = 16,
 col = rgb(0, 0, 0, 0.4), log10((rowData(sce)$detected * dim(sce)[2]) + 1),
 xlab = "log10(Mean Counts Per Gene + 1e-6)", ylab = "log10(Number of Cells Expressing Gene + 1)")
```

# Cluster Pipeline

We run a pipeline that applies normalization and scaling, dimensionality reduction, and clustering.

```{r}
# Remove redundant QC metrics.
# ----------------------------
if (length(mito) != 0) {
  colData(sce)[(length(names(colData(sce))) - 1):length(names(colData(sce)))] <- NULL
  colData(sce)[(length(names(colData(sce))) - 4):(length(names(colData(sce))) - 2)] <- NULL
} else {
  colData(sce)[(length(names(colData(sce))) - 2):length(names(colData(sce)))] <- NULL
}
rowData(sce) <- NULL
# ----------------------------

sub_name <- "all"
seurat <- as.Seurat(sce, data = NULL)
rm(sce)
seurat <- cluster_pipeline(seurat, cache_dir, sub_name, protocol, vars_to_regress, parallel_override, FALSE, 0.8)
seurat

for (i in 1:length(metadata_to_plot)) {
  print(names(metadata_to_plot)[i])
  if ("no_legend" %in% metadata_to_plot[[i]] && "no_label" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i]) +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i], "cat") +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i]) + ggtitle(names(metadata_to_plot)[i]))
  } else {
    print(
      red_dim_plot(seurat, "umap1", "umap2", names(metadata_to_plot)[i], "cat") + ggtitle(names(metadata_to_plot)[i])
    )
  }
}

if (protocol[4] == "umis") {
  red_dim_plot(seurat, "umap1", "umap2", "nFeature_SCT", "cont") + ggtitle("nFeature_SCT")
  red_dim_plot(seurat, "umap1", "umap2", "nCount_SCT", "cont") + ggtitle("nCount_SCT")
} else if (protocol[4] == "reads") {
  red_dim_plot(seurat, "umap1", "umap2", "nFeature_RNA", "cont") + ggtitle("nFeature_RNA")
  red_dim_plot(seurat, "umap1", "umap2", "nCount_RNA", "cont") + ggtitle("nCount_RNA")
}

if (length(mito) != 0) {
  red_dim_plot(seurat, "umap1", "umap2", "subsets_mito_percent", "cont") + ggtitle("subsets_mito_percent")
}
```

# References

This is the concluding section of the document. Here we output the `sessionInfo` and create a bibliography for works cited.

```{r}
sessionInfo()
```
